substrate/primitives/io/src/global_alloc_wasm.rs

```diff
@@ -17,69 +17,177 @@
 
 use core::{
 	alloc::{GlobalAlloc, Layout},
-	ptr,
+	cell::UnsafeCell,
+	ptr::NonNull,
 };
 
-/// The type used to store the offset between the real pointer and the returned pointer.
-type Offset = u16;
-
-/// The length of [`Offset`].
-const OFFSET_LENGTH: usize = core::mem::size_of::<Offset>();
-
 /// Allocator used by Substrate from within the runtime.
-///
-/// The allocator needs to align the returned pointer to given layout. We assume that on the host
-/// side the freeing-bump allocator is used with a fixed alignment of `8` and a `HEADER_SIZE` of
-/// `8`. The freeing-bump allocator is storing the header in the 8 bytes before the actual pointer
-/// returned by `alloc`. The problem is that the runtime not only sees pointers allocated by this
-/// `RuntimeAllocator`, but also pointers allocated by the host. The header is stored as a
-/// little-endian `u64`. The allocation header consists of 8 bytes. The first four bytes (as written
-/// in memory) are used to store the order of the allocation (or the link to the next slot, if
-/// unallocated). Then the least significant bit of the next byte determines whether a given slot is
-/// occupied or free, and the last three bytes are unused.
-///
-/// The `RuntimeAllocator` aligns the pointer to the required alignment before returning it to the
-/// user code. As we are assuming the freeing-bump allocator that already aligns by `8` by default,
-/// we only need to take care of alignments above `8`. The offset is stored in two bytes before the
-/// pointer that we return to the user. Depending on the alignment, we may write into the header,
-/// but given the assumptions above this should be no problem.
 struct RuntimeAllocator;
 
 #[global_allocator]
 static ALLOCATOR: RuntimeAllocator = RuntimeAllocator;
 
+/// The size of the local heap.
+///
+/// This should be as big as possible, but it should still leave enough space
+/// under the maximum memory usage limit to allow the host allocator to service the host calls.
+const LOCAL_HEAP_SIZE: usize = 64 * 1024 * 1024;
+const LOCAL_HEAP_S: picoalloc::Size = picoalloc::Size::from_bytes_usize(LOCAL_HEAP_SIZE).unwrap();
+
+#[repr(align(32))] // `picoalloc` requires 32-byte alignment of the heap
+struct LocalHeap(UnsafeCell<[u8; LOCAL_HEAP_SIZE]>);
+
+// SAFETY: This is runtime-only, and runtimes are single-threaded, so this is safe.
+unsafe impl Send for LocalHeap {}
+
+// SAFETY: This is runtime-only, and runtimes are single-threaded, so this is safe.
+unsafe impl Sync for LocalHeap {}
+
+// Preallocate a bunch of space statically for use by the local allocator.
+//
+// This should be relatively cheap as long as all of this space is full of zeros,
+// since none of this memory will be physically paged-in until it's actually used.
+static LOCAL_HEAP: LocalHeap = LocalHeap(UnsafeCell::new([0; LOCAL_HEAP_SIZE]));
```

Comment:
I didn't realize before that you're getting that heap compiled into your Wasm module. That is, literally, you get a 64 Mb sized global in your on-disk module's `bss` segment filled with zeroes. May pose a problem with real values :sweat_smile:

Reply:
Exactly, and it doesn't get compressed because it's over the code bomb size limit, so you're putting all the 70 Mb on chain without any compression :facepalm:

Reply:
In wasm there is no bss. Just memory and initializers. I assume Rust generates those data initializers with zeroes and nothing in between seems to optimize this out.  They can be removed because the memory is always zero initialized. We are running `wasm-opt` but AFAIK we disable optimizations. You can try to enable them (in wasm-builder) and see if it removes the zeroes.

Reply:
> And if LLVM is dumb enough to initialize 64 Mb zeroed memory from a vector compiled into the blob, I'd doubt if wasm-opt will do any better job optimizing it out. 

I wouldn't be so sure. The precise reason for the existence of wasm-opt is that LLVM is really bad at optimizing wasm. LLVM was never meant to deal with stack machines. Having a wasm backend is a giant hack.

> Indeed, we do not need that; we just need dynamic heap allocation.

We are fine with the static buffer for now.

Reply:
A quick update. By now I've tried:
* `wasm-opt`'s `OptimizeLevel` up to `Level4` (max available)
* `wasm-opt`'s `ShrinkLevel` up to `Level2` (max available)
* Different combinations of `UnsafeCell` and `MaybeUninit`

with no success.